File: api/lib/logger.js
/**
 * Centralized logging utility with emoji labels and color coding
 * Provides consistent logging across all application components
 */

// ANSI color codes for terminal output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
  
  // Foreground colors
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  white: '\x1b[37m',
  
  // Background colors
  bgRed: '\x1b[41m',
  bgGreen: '\x1b[42m',
  bgYellow: '\x1b[43m',
  bgBlue: '\x1b[44m',
  bgMagenta: '\x1b[45m',
  bgCyan: '\x1b[46m',
};

// Component configurations with emoji and color
const componentConfig = {
  BACKGROUND: {
    emoji: '🌄',
    color: colors.blue,
    name: 'BACKGROUND'
  },
  CHARACTER: {
    emoji: '🎭',
    color: colors.magenta,
    name: 'CHARACTER'
  },
  MACRO_CHAIN: {
    emoji: '🔗',
    color: colors.cyan,
    name: 'MACRO_CHAIN'
  },
  SCENE: {
    emoji: '🎬',
    color: colors.green,
    name: 'SCENE'
  },
  CONTEXT: {
    emoji: '🧠',
    color: colors.yellow,
    name: 'CONTEXT'
  },
  STORAGE: {
    emoji: '💾',
    color: colors.white,
    name: 'STORAGE'
  },
  API: {
    emoji: '🌐',
    color: colors.cyan,
    name: 'API'
  },
  AI: {
    emoji: '🤖',
    color: colors.magenta,
    name: 'AI'
  },
  LOCK: {
    emoji: '🔒',
    color: colors.red,
    name: 'LOCK'
  },
  VALIDATION: {
    emoji: '✅',
    color: colors.green,
    name: 'VALIDATION'
  },
  PROMPT: {
    emoji: '📝',
    color: colors.blue,
    name: 'PROMPT'
  },
  SERVER: {
    emoji: '⚙️',
    color: colors.white,
    name: 'SERVER'
  },
  ERROR: {
    emoji: '❌',
    color: colors.red,
    name: 'ERROR'
  }
};

// Log level configurations
const logLevels = {
  DEBUG: { emoji: '🔍', color: colors.dim, name: 'DEBUG' },
  INFO: { emoji: 'ℹ️', color: colors.white, name: 'INFO' },
  WARN: { emoji: '⚠️', color: colors.yellow, name: 'WARN' },
  ERROR: { emoji: '❌', color: colors.red, name: 'ERROR' },
  SUCCESS: { emoji: '✅', color: colors.green, name: 'SUCCESS' }
};

/**
 * Format timestamp for logs
 */
function getTimestamp() {
  const now = new Date();
  return now.toISOString().substring(11, 23); // HH:MM:SS.mmm
}

/**
 * Create a logger for a specific component
 * @param {string} component - Component name from componentConfig
 * @returns {Object} Logger instance with log methods
 */
function createLogger(component) {
  const config = componentConfig[component] || componentConfig.API;
  
  const logger = {
    /**
     * Log debug information
     */
    debug: (...args) => {
      const level = logLevels.DEBUG;
      console.log(
        `${colors.dim}[${getTimestamp()}]${colors.reset}`,
        `${config.emoji} ${config.color}${config.name}${colors.reset}`,
        `${level.emoji} ${level.color}${level.name}${colors.reset}`,
        ...args
      );
    },
    
    /**
     * Log general information
     */
    info: (...args) => {
      const level = logLevels.INFO;
      console.log(
        `${colors.dim}[${getTimestamp()}]${colors.reset}`,
        `${config.emoji} ${config.color}${config.name}${colors.reset}`,
        `${level.emoji}`,
        ...args
      );
    },
    
    /**
     * Log warnings
     */
    warn: (...args) => {
      const level = logLevels.WARN;
      console.warn(
        `${colors.dim}[${getTimestamp()}]${colors.reset}`,
        `${config.emoji} ${config.color}${config.name}${colors.reset}`,
        `${level.emoji} ${level.color}${level.name}${colors.reset}`,
        ...args
      );
    },
    
    /**
     * Log errors
     */
    error: (...args) => {
      const level = logLevels.ERROR;
      console.error(
        `${colors.dim}[${getTimestamp()}]${colors.reset}`,
        `${config.emoji} ${config.color}${config.name}${colors.reset}`,
        `${level.emoji} ${level.color}${level.name}${colors.reset}`,
        ...args
      );
    },
    
    /**
     * Log success messages
     */
    success: (...args) => {
      const level = logLevels.SUCCESS;
      console.log(
        `${colors.dim}[${getTimestamp()}]${colors.reset}`,
        `${config.emoji} ${config.color}${config.name}${colors.reset}`,
        `${level.emoji} ${level.color}${level.name}${colors.reset}`,
        ...args
      );
    },
    
    /**
     * Log with custom emoji and message
     */
    custom: (emoji, ...args) => {
      console.log(
        `${colors.dim}[${getTimestamp()}]${colors.reset}`,
        `${config.emoji} ${config.color}${config.name}${colors.reset}`,
        emoji,
        ...args
      );
    },
    
    /**
     * Create a sub-logger with additional context
     */
    child: (context) => {
      const childLogger = createLogger(component);
      const contextStr = typeof context === 'string' ? context : JSON.stringify(context);
      
      // Wrap all methods to include context
      Object.keys(childLogger).forEach(method => {
        if (typeof childLogger[method] === 'function' && method !== 'child') {
          const originalMethod = childLogger[method];
          childLogger[method] = (...args) => {
            originalMethod(`[${contextStr}]`, ...args);
          };
        }
      });
      
      return childLogger;
    },
    
    /**
     * Log a separator line
     */
    separator: (char = '─', length = 80) => {
      console.log(`${config.color}${char.repeat(length)}${colors.reset}`);
    },
    
    /**
     * Log a section header
     */
    section: (title) => {
      console.log(
        `\n${config.color}${'═'.repeat(80)}${colors.reset}`
      );
      console.log(
        `${config.emoji} ${config.color}${config.name}${colors.reset}`,
        `${colors.bright}${title}${colors.reset}`
      );
      console.log(
        `${config.color}${'═'.repeat(80)}${colors.reset}\n`
      );
    }
  };
  
  return logger;
}

/**
 * Create loggers for all components
 */
const loggers = {
  background: createLogger('BACKGROUND'),
  character: createLogger('CHARACTER'),
  macroChain: createLogger('MACRO_CHAIN'),
  scene: createLogger('SCENE'),
  context: createLogger('CONTEXT'),
  storage: createLogger('STORAGE'),
  api: createLogger('API'),
  ai: createLogger('AI'),
  lock: createLogger('LOCK'),
  validation: createLogger('VALIDATION'),
  prompt: createLogger('PROMPT'),
  server: createLogger('SERVER'),
  error: createLogger('ERROR')
};

/**
 * Global logger - auto-detects component from context if possible
 */
const logger = {
  ...loggers,
  
  /**
   * Create a logger for a specific component
   */
  create: createLogger,
  
  /**
   * Log a grouped set of related messages
   */
  group: (component, title, fn) => {
    const log = loggers[component] || loggers.api;
    log.section(title);
    fn(log);
    log.separator();
  }
};

export default logger;
export { loggers, createLogger };

File: backend/utils/logger.py
"""
Centralized logging utility with emoji labels and color coding
Provides consistent logging across all application components
"""
import sys
from datetime import datetime
from typing import Any, Optional
from enum import Enum


class LogLevel(Enum):
    """Log levels"""
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARN = "WARN"
    ERROR = "ERROR"
    SUCCESS = "SUCCESS"


# ANSI color codes for terminal output
class Colors:
    RESET = '\x1b[0m'
    BRIGHT = '\x1b[1m'
    DIM = '\x1b[2m'
    
    # Foreground colors
    RED = '\x1b[31m'
    GREEN = '\x1b[32m'
    YELLOW = '\x1b[33m'
    BLUE = '\x1b[34m'
    MAGENTA = '\x1b[35m'
    CYAN = '\x1b[36m'
    WHITE = '\x1b[37m'


# Component configurations with emoji and color
COMPONENT_CONFIG = {
    'BACKGROUND': {'emoji': '🌄', 'color': Colors.BLUE, 'name': 'BACKGROUND'},
    'CHARACTER': {'emoji': '🎭', 'color': Colors.MAGENTA, 'name': 'CHARACTER'},
    'MACRO_CHAIN': {'emoji': '🔗', 'color': Colors.CYAN, 'name': 'MACRO_CHAIN'},
    'SCENE': {'emoji': '🎬', 'color': Colors.GREEN, 'name': 'SCENE'},
    'CONTEXT': {'emoji': '🧠', 'color': Colors.YELLOW, 'name': 'CONTEXT'},
    'STORAGE': {'emoji': '💾', 'color': Colors.WHITE, 'name': 'STORAGE'},
    'API': {'emoji': '🌐', 'color': Colors.CYAN, 'name': 'API'},
    'AI': {'emoji': '🤖', 'color': Colors.MAGENTA, 'name': 'AI'},
    'LOCK': {'emoji': '🔒', 'color': Colors.RED, 'name': 'LOCK'},
    'VALIDATION': {'emoji': '✅', 'color': Colors.GREEN, 'name': 'VALIDATION'},
    'PROMPT': {'emoji': '📝', 'color': Colors.BLUE, 'name': 'PROMPT'},
    'SERVER': {'emoji': '⚙️', 'color': Colors.WHITE, 'name': 'SERVER'},
    'ERROR': {'emoji': '❌', 'color': Colors.RED, 'name': 'ERROR'},
}

# Log level configurations
LOG_LEVELS = {
    LogLevel.DEBUG: {'emoji': '🔍', 'color': Colors.DIM, 'name': 'DEBUG'},
    LogLevel.INFO: {'emoji': 'ℹ️', 'color': Colors.WHITE, 'name': 'INFO'},
    LogLevel.WARN: {'emoji': '⚠️', 'color': Colors.YELLOW, 'name': 'WARN'},
    LogLevel.ERROR: {'emoji': '❌', 'color': Colors.RED, 'name': 'ERROR'},
    LogLevel.SUCCESS: {'emoji': '✅', 'color': Colors.GREEN, 'name': 'SUCCESS'},
}


def get_timestamp() -> str:
    """Format timestamp for logs"""
    now = datetime.now()
    return now.strftime("%H:%M:%S.%f")[:-3]


def create_logger(component: str):
    """Create a logger for a specific component"""
    config = COMPONENT_CONFIG.get(component, COMPONENT_CONFIG['API'])
    
    class Logger:
        def __init__(self, component_name: str):
            self.component = component_name
            self.config = COMPONENT_CONFIG.get(component_name, COMPONENT_CONFIG['API'])
        
        def _log(self, level: LogLevel, *args: Any, use_stderr: bool = False):
            """Internal log method"""
            level_config = LOG_LEVELS[level]
            timestamp = get_timestamp()
            
            output = sys.stderr if use_stderr else sys.stdout
            
            print(
                f"{Colors.DIM}[{timestamp}]{Colors.RESET}",
                f"{self.config['emoji']} {self.config['color']}{self.config['name']}{Colors.RESET}",
                f"{level_config['emoji']} {level_config['color']}{level_config['name']}{Colors.RESET}",
                *args,
                file=output
            )
        
        def debug(self, *args: Any):
            """Log debug information"""
            self._log(LogLevel.DEBUG, *args)
        
        def info(self, *args: Any):
            """Log general information"""
            self._log(LogLevel.INFO, *args)
        
        def warn(self, *args: Any):
            """Log warnings"""
            self._log(LogLevel.WARN, *args, use_stderr=True)
        
        def error(self, *args: Any):
            """Log errors"""
            self._log(LogLevel.ERROR, *args, use_stderr=True)
        
        def success(self, *args: Any):
            """Log success messages"""
            self._log(LogLevel.SUCCESS, *args)
        
        def custom(self, emoji: str, *args: Any):
            """Log with custom emoji and message"""
            timestamp = get_timestamp()
            print(
                f"{Colors.DIM}[{timestamp}]{Colors.RESET}",
                f"{self.config['emoji']} {self.config['color']}{self.config['name']}{Colors.RESET}",
                emoji,
                *args
            )
        
        def separator(self, char: str = '─', length: int = 80):
            """Log a separator line"""
            print(f"{self.config['color']}{char * length}{Colors.RESET}")
        
        def section(self, title: str):
            """Log a section header"""
            print(f"\n{self.config['color']}{'═' * 80}{Colors.RESET}")
            print(
                f"{self.config['emoji']} {self.config['color']}{self.config['name']}{Colors.RESET}",
                f"{Colors.BRIGHT}{title}{Colors.RESET}"
            )
            print(f"{self.config['color']}{'═' * 80}{Colors.RESET}\n")
    
    return Logger(component)


# Create loggers for all components
loggers = {
    'background': create_logger('BACKGROUND'),
    'character': create_logger('CHARACTER'),
    'macroChain': create_logger('MACRO_CHAIN'),
    'scene': create_logger('SCENE'),
    'context': create_logger('CONTEXT'),
    'storage': create_logger('STORAGE'),
    'api': create_logger('API'),
    'ai': create_logger('AI'),
    'lock': create_logger('LOCK'),
    'validation': create_logger('VALIDATION'),
    'prompt': create_logger('PROMPT'),
    'server': create_logger('SERVER'),
    'error': create_logger('ERROR'),
}

# Export default logger
logger = loggers

